include/gridtools/storage/README.md

```diff
@@ -0,0 +1,252 @@
+    unsigned length() const;
+    
+    // Supplementary object that holds lengths and strides. 
+    storage::info<ndims> const& info() const;
+
+    // Request the target view.
+    // If the target and host spaces are different necessary synchronization is performed
+    // and the host counterpart is marked as dirty.  
+    auto target_view();
+    // Const version doesn't mark host counterpart as dirty. Synchronization takes place. 
+    auto const_target_view();
+
+    // Raw ptr alternatives for target_view/const_target_view.
+    // Synchronization behaviour is the same.
+    data_t *get_target_ptr();
+    data_t const *get_const_target_ptr();
+
+    // Host access methods variations. They only exist if !std::is_const<data_t>::value.  
+    auto host_view();
+    auto host_const_view();
+    data_t *get_host_ptr();
+    data_t const *get_const_host_ptr();
+};
+ ```
+
+### Data View Synopsis
+
+Data view is a supplemental struct that is returned form data store access methods. The distinctive property:
+data view is a POD. Hence it can be passed to the target device by copying the memory. For the cuda data stores
+all data view methods are declared as device only.
+
+```C++
+template <class T, size_t N>
+struct some_view {
+    // POD members here
+
+    // The meta info methods are the same as for data_store. 
+    array<unsigned, N> const& lengths() const;
+    array<unsigned, N> const& strides() const;
+    unsigned length() const;
+    storage::info<N> const& info() const;
+
+    // raw access
+    T *data() const;
+    // multi dimensional indexed access
+    T &operator()(int... /*number of arguments equals to N*/ ) const;
+    // variation with array as an argument
+    T& operator()(array<int, N> const &) const;
+};
+```
+
+## Builder API
+
+The builder design pattern is used for data store construction. The API is defined in [builder.hpp](builder.hpp).
+Here a single user facing symbol is defined -- `storage::builder`.
+It is a value template parametrized by `Traits` (see below).
+The idea is that the user takes a builder with the desired traits, customize it with requested properties and finally
+calls `build()` method (or alternatively overloaded call operator) to produce `std::shared_ptr` to a data store.
+For example:
+```C++
+auto ds = storage::builder<storage::cuda>
+        .type<double>()
+        .name("my special data")
+        .dimensions(132, 132, 80)
+        .halo(2, 2, 0)
+        .selector<1, 1, 0>()
+        .value(42)
+        .build();
+
+assert(ds->const_host_view()(1, 2, 3) == 42);
+
+```
+One can also use partially specified builder to produce several data stores:
+```C++
+auto const my_builder = storage::builder<storage::cuda>.dimensions(10, 10, 10);
+auto foo = my_builder.type<int>().name("foo")();
+auto bar = my_builder.type<tuple<int, double>>()();
+auto baz = my_builder.type<double const>.initialize([](int i, int j, int k){ return i + j + k; })();
```

Comment:
Is there a way to cast a non-const data store into a const one? In the dycore, constant fields are often initialized by GT stencils on the GPU, but afterwards used as const-only, so it might make sense to support this use case.

Reply:
I think `sid::as_const` can be used in this case